### ISSUE TITLE
Code restructure testing / input / code review

### ISSUE THREAD

**Issue by SweVictor** [icanos/hassio-plejd#158]
I've been working at restructuring the code-base for a while. At least to me, the flow is quite hard to follow with `setTimeout`, `Promise` and `Event` that creates quite many moving parts.

I have now published the initial restructure in a new branch and would be happy for any input. Major changes include
* break out main logic into `PlejdAddon.js`
* create central `DeviceRegistry` as a central state store to avoid passing around loose lists/dictionaries
* start rewriting setTimout-based code fragments to a stricter async/await model
* improve logging
* rename `PlejdService.js` to `PlejdBLE.js`

As far as I can see (with limited testing) the reworked code works the same or better than the most recent `0.5.1` version. I plan to continue work from here but would very much appreciate **code review, general input, and testing**.

Branch is here
https://github.com/SweVictor/hassio-plejd/tree/features/code-architecture

Diff to master here
https://github.com/icanos/hassio-plejd/compare/master...SweVictor:features/code-architecture

Thanks!

**Comment by SweVictor** [comment]
Update: "Code works the same or better" is a matter of opinion. Code-wise I still think it's true, longer term since the addon doesn't crash anymore it is more evident that event subscriptions aren't cleaned up properly.

Also there is the outstanding question on how to handle errors in for example the BLE write loop...

**Comment by icanos** [comment]
Not sure how to handle write errors, should we just throw them away? The device state won't change in HA since the device won't be reporting back any new values.

**Comment by SweVictor** [comment]
It's a lot more stable since the last few commits (pushed the most recent code just now).

I'm still not 100% finished with event subscription handling (they should probably be unsubscribed during reconnect attempts, or potentially re-used). I also had the idea to change the BLE to a more clear state machine to not have to restart the ENTIRE ble init loop when something happens. But anyway, it's a lot more stable now and at least to me the code is a lot easier to read.

I must say though... After wrestling bluez and dbus for quite some time I'm really impressed you figured out how to get the comm channel up and running in the first place! 😃    Commands neet to happen in the right order, async/callback-based after waiting enough time during discovery etc. Really good job!

**Comment by SweVictor** [comment]
PR in! Code deemed stable by ... well ... me ... now 😃 
Please have a look!

**Comment by SweVictor** [comment]
Merged to [develop branch](https://github.com/icanos/hassio-plejd/tree/develop), planned for 0.6.0 release. Closing
